lms/djangoapps/ccx/views.py

```diff
@@ -696,56 +707,41 @@ def ccx_gradebook(request, course, ccx=None):
-    Download grades as CSV.
+    List grade book CSV files that are available for download.
     """
     if not ccx:
         raise Http404
 
     ccx_key = CCXLocator.from_course_locator(course.id, ccx.id)
-    with ccx_course(ccx_key) as course:
-        prep_course_for_grading(course, request)
-
-        enrolled_students = User.objects.filter(
-            courseenrollment__course_id=ccx_key,
-            courseenrollment__is_active=1
-        ).order_by('username').select_related("profile")
-        grades = iterate_grades_for(course, enrolled_students)
-
-        header = None
-        rows = []
-        for student, gradeset, __ in grades:
-            if gradeset:
-                # We were able to successfully grade this student for this
-                # course.
-                if not header:
-                    # Encode the header row in utf-8 encoding in case there are
-                    # unicode characters
-                    header = [section['label'].encode('utf-8')
-                              for section in gradeset[u'section_breakdown']]
-                    rows.append(["id", "email", "username", "grade"] + header)
-
-                percents = {
-                    section['label']: section.get('percent', 0.0)
-                    for section in gradeset[u'section_breakdown']
-                    if 'label' in section
-                }
-
-                row_percents = [percents.get(label, 0.0) for label in header]
-                rows.append([student.id, student.email, student.username,
-                             gradeset['percent']] + row_percents)
+    return list_report_downloads_by_course_key(ccx_key)
 
-        buf = StringIO()
-        writer = csv.writer(buf)
-        for row in rows:
-            writer.writerow(row)
 
-        response = HttpResponse(buf.getvalue(), content_type='text/csv')
-        response['Content-Disposition'] = 'attachment'
+# Grades can potentially be written - if so, let grading manage the transaction.
+@transaction.non_atomic_requests
+@cache_control(no_cache=True, no_store=True, must_revalidate=True)
+@coach_dashboard
+def ccx_grades_csv(request, course, ccx=None):
+    """
+    Download grades as CSV.
+    """
+    if not ccx:
+        raise Http404
 
-        return response
+    ccx_key = CCXLocator.from_course_locator(course.id, ccx.id)
+    try:
+        submit_calculate_grades_csv(request, ccx_key)
+        success_status = _("The grade book is being created."
+                           " Grade book will shortly appear under 'Available for Download' section.")
+        return JsonResponse({"status": success_status})
+    except AlreadyRunningError:
+        already_running_status = _("The grade report is currently being created."
+                                   " To view the status of the report, see Pending Instructor Tasks below."
+                                   " You will be able to download the report when it is complete.")
+        return JsonResponse({
```

Comment:
This seems like it should be a 400 Bad Request status code, rather than the default 200 OK.